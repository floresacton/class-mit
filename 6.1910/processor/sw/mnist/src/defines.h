#ifndef DEFINES_H_
#define DEFINES_H_

// Typedefs to make the bit size easier to interpret
typedef signed char int8_t;
typedef unsigned char uint8_t;
typedef signed short int16_t;
typedef unsigned short uint16_t;
typedef signed int int32_t;
typedef unsigned int uint32_t;
typedef long long int64_t;
typedef unsigned long long uint64_t;

#endif