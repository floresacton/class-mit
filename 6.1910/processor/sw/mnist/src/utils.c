#include "utils.h"

void print_string(char* s) {
    while (*s != '\0') {
        print_char(*s);
        s++;
    }
}

const int INT_PRINT_LIMIT = 10;
void print_int(int a) {
    int bases[INT_PRINT_LIMIT];
    bases[0] = 1;
    int i = 0;
    for (i = 1; i < INT_PRINT_LIMIT; i++) {
        bases[i] = bases[i - 1] * 10;
    }
    i = 0;
    if (a == 0) {
        print_char('0');
    } else if (a == 0x80000000) {
        print_string("-2147483648");
    } else {
        if (a < 0) {
            print_char('-');
            a = -a;
        }
        int j = INT_PRINT_LIMIT - 1;
        while (a < bases[j]) j--;
        while (j >= 0) {
            int d = 0;
            while (a >= bases[j]) {
                a -= bases[j];
                d ++;
            }
            print_char('0' + d);
            j--;
        }
    }
}

void print_int_hex(unsigned int a) {
    const char hex_lookup[] = "0123456789abcdef";
    int len = 0;

    if (a == 0) len = 1;
    else for (unsigned int n = a; n; n >>= 4) len++;

    for (int i = len - 1; i >= 0; i--)
        print_char(hex_lookup[(a >> (4 * i)) & 0xf]);
}