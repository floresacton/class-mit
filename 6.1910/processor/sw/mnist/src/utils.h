#ifndef UTILS_H_
#define UTILS_H_


// this is linked  from the crt0.S
void print_char(char x);

void print_string(char* s);

void print_int(int a);

void print_int_hex(unsigned int a);

#endif