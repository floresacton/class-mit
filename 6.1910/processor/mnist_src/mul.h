#ifndef MUL_H_
#define MUL_H_

#include "defines.h"

// GCC uses this unsigned multiplication function to override the * operator for unsigned/signed int multiply
uint32_t __mulsi3 (uint32_t a, uint32_t b) {
    uint32_t result = 0;

    // Examples of how to inline new instructions into C
    // read more at : https://sourceware.org/binutils/docs-2.34/as/RISC_002dV_002dFormats.html
    // and  https://gcc.gnu.org/onlinedocs/gcc/Extended-Asm.html
    //
    // asm volatile (".insn r 0b0110011, 0b000, 0b0000001, %0, %1, %2" : "=r" (result) : "r"(a), "r"(b)); 
 
    while (b) {
        if (b & 1) {
            result += a;
        }
        a = a << 1;
        b = b >> 1;
    }
    return result;

}

// Same as above but for 64 bit multiplication

// WARNING: Using unsigned 32x32 bit multiplication here will cause signed mulitplicaitons
// to fail! Since only 64x64 bit multiplicaiton will work, we suggest you implement and
// call a separate helper function instead which computes signed 32x32 bit multiplicaiton directly.

// If you would like to test your function with the provided testbench,
// make sure to change call to * in sw/mnist/mul64_test.c
uint64_t __muldi3 (uint64_t a, uint64_t b) {
    uint64_t result = 0;
    
    while (b) {
        if (b & 1) {
            result += a;
        }
        a = a << 1;
        b = b >> 1;
    }
    return result;

}

// SUGGESTED: Integrate into model.c and then implement in hardware
int32_t packmul(uint32_t input_i, uint32_t filter_i) {
    // input_i and filter_i each hold four adjacent int8_t numbers cast to a single uint32_t

    // NOTE: Packed elements of input buffer should be treated as unsigned.
    uint8_t a[4]; *(uint32_t*)a = input_i;
    int8_t b[4]; *(uint32_t*)b = filter_i; 

    // NOTE: software implementation of packmul relies on working 32 bit multiplication
    return (
        ((int32_t)a[0] * (int32_t)b[0]) +
        ((int32_t)a[1] * (int32_t)b[1]) +
        ((int32_t)a[2] * (int32_t)b[2]) +
        ((int32_t)a[3] * (int32_t)b[3])
    );
}

#endif