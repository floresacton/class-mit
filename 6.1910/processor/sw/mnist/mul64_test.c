#include "src/defines.h"
#include "src/utils.h"
#include "src/mul.h"

// const volatile prevents constant folding
const volatile int32_t a[] = {1484724050,-1630402562,-2043333423,580705473,183380344,1803096246,-1712203173,1581898420,-512835935,-1164803001,478265090,406173370,462765880,-1852980333,-1572526809,801844850,-788462765,-762106559,645675011,-1376974680,-1257025100,1186470642,171178518,1112967154,-310500125,1011806184,-2116191498,-1263463804,-824624764,1461336268,354530567,-746581126,634800978,-397444909,-636079657,-1923653799,-1720712074,-925680499,26802786,-386300563,579380936,1251971273,1945771283,110570954,848567295,-2011129181,-1718946899,905317097,1359987653,-1505302564,-780263095,27836977,-913754740,-1755691792,2015110632,-708388526,-158457525,1390832814,-1978119939,1449646426,1700059371,-48917702,1651062619,-1582491102,-1945184376,-2077438537,-1473856563,1876598414,-1104938533,155069376,-276185420,1090179216,-49484201,1264013912,-1270982314,1451581984,1451372186,-1780926368,-1910157566,1503684194,2047841183,989067278,1525531616,-1708422475,-1618235223,1423212260,-1469700850,1344951546,-1759360100,1707541109,1351496797,1060125183,-1691460878,1551230368,-34264088,73435908,-1983704527,-1001701297,714063198,-546243739,-229712277,1564749263,2066309952,1515338540,1071498698,-1178603236,790344107,641898971,-207860686,1816519917,1025462755,-1958866988,-162973861,-156010488,-2052408055,-861413709,1983741206,284334959,189514083,1459843392,1353072445,-2114310531,-615986371,-2090007421,-1259806529,-860868340,152064950,-781769816,-1543666243,-316998861,391963344,952352964,749483846,-967720287,272349094,-2135336622,1165285479,-601085179,-1916591929,-19785971,842326895,-6287342,-1384974166,-900106619,-227812224,1759730961,-1775154610,-2062738044,1623149186,-1512317297,146264302,-821920904,21945294,-1192435988,237409443,-1379145243,1639913352,1525528563,181653642,319712796,-2074576469,-1383524532,1807308841,-1527343832,524318596,1488670221,747598074,246114683,1416233495,1003045139,34097309,855718216,2020576483,1730366039,-391585211,-1574008061,-944818531,1009311449,1348873212,1074161569,939492001,-34167178,727043879,86364705,-503480467,-750285791,2068205098,1022633640,-267245308,2077825339,-227832924,-1485331162,-1399865976,-2138792749,631991117,1185131674,1203334916,-1369068157,259505549,653863124};
const volatile int32_t b[] = {1308686729,-128862990,1242481624,-1428976981,-1721306711,-1807117456,-1158551715,-1228168787,504722542,111239084,-879534830,-1266756570,-2110682403,-533669819,1458020677,-1629568558,1945932600,-1464397170,-2010941350,1478608788,-69117077,1475789333,919300768,-263565905,-1286502645,641869746,-1181859328,-2115092705,1494098458,1345569517,-430682231,-67391069,1198866242,678289909,-154231996,-185034486,-780624305,625384110,-399503713,-1109472155,-1499554105,-731706636,845634204,-1301753444,1822436063,1502479642,603053296,-607057222,-686449935,-1406296350,-941121511,-753808406,-350637837,1044197688,-1215473579,-1549413010,-1102610057,-1430982481,-1566829474,-1877025081,1244030863,-38482958,-476959065,-1374740623,-409604244,-1809021869,-419601403,1813845196,-2062128359,1510795705,-1019500733,1553430442,-1062740974,-1650098322,-1916399313,36072044,-1308315160,748559907,2061580447,-12939674,1974135899,564463273,355175675,1628592447,-1243867976,-1067073325,445009960,1907772793,-963191270,1385387107,-811971749,1657446035,248402621,1531696677,512076672,1211530765,-1268161814,-204058113,1244701187,1263524865,1551680796,-724597396,1120315128,890487819,-144767400,-1952842081,-763953030,210785020,-738178219,-1708997924,-1689322140,1019369391,901574707,-1546160339,684299935,-410114037,-1018464230,-460759503,-2109411868,-1218237294,682712340,-1915071551,-1369823592,-169899196,1378280051,1105042007,1042671604,-1122429636,452848160,-625893329,-187694308,1065889554,-1529970094,-1009636869,-1889268247,-469861053,-799253821,807206673,-1249392042,-82003652,-1995831215,-1559609756,-1006950049,1127868568,1607603246,-1074528372,-397048461,-39161783,-1048786364,1964878361,-1592313576,-1452323621,-120737430,-862137511,-2006913454,-2115902633,-984658232,2067813789,1609760137,-383503523,1911074819,1796262786,1822349377,107667676,-1599424649,579912250,1137527449,-1878460116,546582163,1812063097,-999465972,-218326864,-2007910601,1480495707,-1977327870,-1617405295,-156268262,1346895014,-161999180,-255589270,838991311,-56172954,-2042919187,1499464039,725395779,-293181479,-1489814648,733201444,2082137598,-780979358,-264611492,1031539858,613598484,-1105127231,-1076525084,1712280238,988943901,-1893553730,1260385011,70633016};

int main(int argc, char* argv[]) {
    for(int test_no = 0; test_no < sizeof(a)/sizeof(a[0]); test_no ++)  {
        print_string("Input ");
        print_int(test_no);
        print_string(" - ");
        print_int(a[test_no]);
        print_string(" * ");
        print_int(b[test_no]);
        print_string(" = 0x");
        int64_t product = (int64_t)a[test_no] * (int64_t)b[test_no];
        print_int_hex((product >> 32) & (uint32_t)-1);
        print_int_hex(product & (uint32_t)-1);
        print_string("\n");

    }
    return 0;
}
