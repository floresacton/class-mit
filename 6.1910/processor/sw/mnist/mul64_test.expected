Input 0 - 1484724050 * 1308686729 = 0x1af70f5df5b23ce2
Input 1 - -1630402562 * -128862990 = 0x2ea6b846788261c
Input 2 - -2043333423 * 1242481624 = 0xdcc45b01118dc358
Input 3 - 580705473 * -1428976981 = 0xf47be7e0cada7eeb
Input 4 - 183380344 * -1721306711 = 0xfb9e9286aa135838
Input 5 - 1803096246 * -1807117456 = 0xd2c7d05e8575d1a0
Input 6 - -1712203173 * -1158551715 = 0x1b876ebdd68f700f
Input 7 - 1581898420 * -1228168787 = 0xe509a6e48f2fffa4
Input 8 - -512835935 * 504722542 = 0xfc686a858d2bdd2e
Input 9 - -1164803001 * 111239084 = 0xfe33ab4c942e38b4
Input 10 - 478265090 * -879534830 = 0xfa298c5267a92824
Input 11 - 406173370 * -1266756570 = 0xf8dc0c40cad7279c
Input 12 - 462765880 * -2110682403 = 0xf271e169cda3c358
Input 13 - -1852980333 * -533669819 = 0xdb934d423c9be9f
Input 14 - -1572526809 * 1458020677 = 0xe02e6bd3d9318a83
Input 15 - 801844850 * -1629568558 = 0xedddce9a58c0b84
Input 16 - -788462765 * 1945932600 = 0xeab51676dbd56328
Input 17 - -762106559 * -1464397170 = 0xf7cec5c2848ac0e
Input 18 - 645675011 * -2010941350 = 0xedfb1ad2256190e
Input 19 - -1376974680 * 1478608788 = 0xe3bea68bac6b9920
Input 20 - -1257025100 * -69117077 = 0x134aaa0d852223c
Input 21 - 1186470642 * 1475789333 = 0x184cbbaf741199da
Input 22 - 171178518 * 919300768 = 0x22f1237d97b69c0
Input 23 - 1112967154 * -263565905 = 0xfbedd8a58bfeb86e
Input 24 - -310500125 * -1286502645 = 0x58b2a0e1eb24ac1
Input 25 - 1011806184 * 641869746 = 0x9034d503f3f2750
Input 26 - -2116191498 * -1181859328 = 0x22b57b3b14276c00
Input 27 - -1263463804 * -2115092705 = 0x251611de2fb7b9fc
Input 28 - -824624764 * 1494098458 = 0xeee6ce53e4eb5f68
Input 29 - 1461336268 * 1345569517 = 0x1b49ce4b92f52cdc
Input 30 - 354530567 * -430682231 = 0xfde1893d663309bf
Input 31 - -746581126 * -67391069 = 0xb2bf50eea1c0ae
Input 32 - 634800978 * 1198866242 = 0xa8fc32da8429924
Input 33 - -397444909 * 678289909 = 0xfc423fce971714ef
Input 34 - -636079657 * -154231996 = 0x15c88eb13146a1c
Input 35 - -1923653799 * -185034486 = 0x4f08fa0ef9dea7a
Input 36 - -1720712074 * -780624305 = 0x12a41c43e9a2346a
Input 37 - -925680499 * 625384110 = 0xf7f7502c818df9d6
Input 38 - 26802786 * -399503713 = 0xffd9f54d1e14dede
Input 39 - -386300563 * -1109472155 = 0x5f2a813ee4a8001
Input 40 - 579380936 * -1499554105 = 0xf3f15b1fce991578
Input 41 - 1251971273 * -731706636 = 0xf3497203cd78d994
Input 42 - 1945771283 * 845634204 = 0x16d5ac5c3476594
Input 43 - 110570954 * -1301753444 = 0xfe00a2dc7e91d118
Input 44 - 848567295 * 1822436063 = 0x157620d96efc3921
Input 45 - -2011129181 * 1502479642 = 0xd610d574e7bee38e
Input 46 - -1718946899 * 603053296 = 0xf19d32b3688a1e30
Input 47 - 905317097 * -607057222 = 0xf85f808ec6a074a
Input 48 - 1359987653 * -686449935 = 0xf30b51004dc4ce75
Input 49 - -1505302564 * -1406296350 = 0x1d60beb31f5e7838
Input 50 - -780263095 * -941121511 = 0xa30d65111fde021
Input 51 - 27836977 * -753808406 = 0xffb573649e7f5bca
Input 52 - -913754740 * -350637837 = 0x472475c60ba13e4
Input 53 - -1755691792 * 1044197688 = 0xe68ed915f0210480
Input 54 - 2015110632 * -1215473579 = 0xde024a2347285608
Input 55 - -708388526 * -1549413010 = 0xf3b6903961c9b3c
Input 56 - -158457525 * -1102610057 = 0x26cb80f20a544dd
Input 57 - 1390832814 * -1430982481 = 0xe4612f72a7bf6af2
Input 58 - -1978119939 * -1566829474 = 0x2b03329e2eca90e6
Input 59 - 1449646426 * -1877025081 = 0xda3cfc4551c200f6
Input 60 - 1700059371 * 1244030863 = 0x1d59ba49de8db845
Input 61 - -48917702 * -38482958 = 0x6b01f30ec2ad4
Input 62 - 1651062619 * -476959065 = 0xf51246ab5d7b765d
Input 63 - -1582491102 * -1374740623 = 0x1e30fb2ff790a702
Input 64 - -1945184376 * -409604244 = 0xb0ea5281ba2e560
Input 65 - -2077438537 * -1809021869 = 0x342790e91c0d0e55
Input 66 - -1473856563 * -419601403 = 0x8951ce231b6bf01
Input 67 - 1876598414 * 1813845196 = 0x2f3cefbe14af7128
Input 68 - -1104938533 * -2062128359 = 0x1f9ef27ddac3ef63
Input 69 - 155069376 * 1510795705 = 0x34052bbd0bf5dc0
Input 70 - -276185420 * -1019500733 = 0x3e8578c9bf0551c
Input 71 - 1090179216 * 1553430442 = 0x1780954482716fa0
Input 72 - -49484201 * -1062740974 = 0xbad550a27c861e
Input 73 - 1264013912 * -1650098322 = 0xe30deff0bc1709d0
Input 74 - -1270982314 * -1916399313 = 0x21cd6101b1e51cca
Input 75 - 1451581984 * 36072044 = 0xba06888ba6a580
Input 76 - 1451372186 * -1308315160 = 0xe5a5ebf4369ab590
Input 77 - -1780926368 * 748559907 = 0xed7fc5423c8df520
Input 78 - -1910157566 * 2061580447 = 0xc9599d42d736823e
Input 79 - 1503684194 * -12939674 = 0xffbadfcb7ee7270c
Input 80 - 2047841183 * 1974135899 = 0x381a9d8172360d85
Input 81 - 989067278 * 564463273 = 0x7bf73c151e5153e
Input 82 - 1525531616 * 355175675 = 0x784f91474dd8ca0
Input 83 - -1708422475 * 1628592447 = 0xd963331e69aa0f8b
Input 84 - -1618235223 * -1243867976 = 0x1bef277be4702d78
Input 85 - 1423212260 * -1067073325 = 0xeaec98025a4395ec
Input 86 - -1469700850 * 445009960 = 0xf6ec69cdcde2a30
Input 87 - 1344951546 * 1907772793 = 0x239bc5dd634d5c2a
Input 88 - -1759360100 * -963191270 = 0x17846dfcc74f5d8
Input 89 - 1707541109 * 1385387107 = 0x20d4519c8bf3573f
Input 90 - 1351496797 * -811971749 = 0xf0c5553bfed5dc0f
Input 91 - 1060125183 * 1657446035 = 0x186279656db2af6d
Input 92 - -1691460878 * 248402621 = 0xfa2b47b0d516a8aa
Input 93 - 1551230368 * 1531696677 = 0x20f94c8210ddf020
Input 94 - -34264088 * 512076672 = 0xffc1aa26be71ec00
Input 95 - 73435908 * 1211530765 = 0x13c159e302b0734
Input 96 - -1983704527 * -1268161814 = 0x22e969ec835e62ca
Input 97 - -1001701297 * -204058113 = 0x2d6318415580db1
Input 98 - 714063198 * 1244701187 = 0xc55a2a02324b41a
Input 99 - -546243739 * 1263524865 = 0xf66bf18922a6c765
Input 100 - -229712277 * 1551680796 = 0xfb0dab97619caab4
Input 101 - 1564749263 * -724597396 = 0xf043e2dd3af0de54
Input 102 - 2066309952 * 1120315128 = 0x20203ded69f3a600
Input 103 - 1515338540 * 890487819 = 0x12b9ff85571a3ae4
Input 104 - 1071498698 * -144767400 = 0xfdd8e8ed72d4b970
Input 105 - -1178603236 * -1952842081 = 0x1ff104a61d2c8864
Input 106 - 790344107 * -763953030 = 0xf79eec094fafab7e
Input 107 - 641898971 * 210785020 = 0x1e0b113ce079994
Input 108 - -207860686 * -738178219 = 0x2211f421537f29a
Input 109 - 1816519917 * -1708997924 = 0xd4eada7bfd0655ac
Input 110 - 1025462755 * -1689322140 = 0xe7f580bc385353ac
Input 111 - -1958866988 * 1019369391 = 0xe449e8bff575ddec
Input 112 - -162973861 * 901574707 = 0xfdf5fd1d324f5721
Input 113 - -156010488 * -1546160339 = 0x358f9cab10fd168
Input 114 - -2052408055 * 684299935 = 0xec82589de3dfae97
Input 115 - -861413709 * -410114037 = 0x4e71856124c89b1
Input 116 - 1983741206 * -1018464230 = 0xe3f634ae3c53043c
Input 117 - 284334959 * -460759503 = 0xfe2e8f115de9823f
Input 118 - 189514083 * -2109411868 = 0xfa73c16fd5cce32c
Input 119 - 1459843392 * -1218237294 = 0xe751ba2fc32d9680
Input 120 - 1353072445 * 682712340 = 0xcd1da25f32319c4
Input 121 - -2114310531 * -1915071551 = 0x383122ee8a21933d
Input 122 - -615986371 * -1369823592 = 0xbb5c0f4ead5ba38
Input 123 - -2090007421 * -169899196 = 0x4ed8900678553cc
Input 124 - -1259806529 * 1378280051 = 0xe7e73022655f93cd
Input 125 - -860868340 * 1105042007 = 0xf2cc519ef0bde314
Input 126 - 152064950 * 1042671604 = 0x2334bd88c670378
Input 127 - -781769816 * -1122429636 = 0xc2d70e0d812360
Input 128 - -1543666243 * 452848160 = 0xf64c7cfc99b879a0
Input 129 - -316998861 * -625893329 = 0x2c0e28b8136fe5d
Input 130 - 391963344 * -187694308 = 0xfefaa11f2dd13ec0
Input 131 - 952352964 * 1065889554 = 0xe165ede924789c8
Input 132 - 749483846 * -1529970094 = 0xf0162575158f406c
Input 133 - -967720287 * -1009636869 = 0xd8f2a3bd5679cdb
Input 134 - 272349094 * -1889268247 = 0xf8dbfc2545859416
Input 135 - -2135336622 * -469861053 = 0xdec7a81867b6a76
Input 136 - 1165285479 * -799253821 = 0xf313260a71190c75
Input 137 - -601085179 * 807206673 = 0xf94439334254d655
Input 138 - -1916591929 * -1249392042 = 0x213b3d01935b08da
Input 139 - -19785971 * -82003652 = 0x5c3acd051cc0c
Input 140 - 842326895 * -1995831215 = 0xe8ab61f9d32d6e1f
Input 141 - -6287342 * -1559609756 = 0x22d652b51ea308
Input 142 - -1384974166 * -1006950049 = 0x135a9d22142b8116
Input 143 - -900106619 * 1127868568 = 0xf1e947316b23e6f8
Input 144 - -227812224 * 1607603246 = 0xfaeae23c5d9b9f00
Input 145 - 1759730961 * -1074528372 = 0xe5c23de3dc2dd44c
Input 146 - -1775154610 * -397048461 = 0x9c8083e6fa4e70a
Input 147 - -2062738044 * -39161783 = 0x11efd6fbb4f2aa4
Input 148 - 1623149186 * -1048786364 = 0xe86015c2bc925e88
Input 149 - -1512317297 * 1964878361 = 0xd6c30abac576b1f7
Input 150 - 146264302 * -1592313576 = 0xfcc493ed32229450
Input 151 - -821920904 * -1452323621 = 0x1090db6894cf3fa8
Input 152 - 21945294 * -120737430 = 0xfff6962fa496714c
Input 153 - -1192435988 * -862137511 = 0xe4458626de6040c
Input 154 - 237409443 * -2006913454 = 0xf96345f7bbab2636
Input 155 - -1379145243 * -2115902633 = 0x287f4e2fd7f78bd3
Input 156 - 1639913352 * -984658232 = 0xe9973da0816dd240
Input 157 - 1525528563 * 2067813789 = 0x2bc711386f9d7f07
Input 158 - 181653642 * 1609760137 = 0x40ee15778e6fbda
Input 159 - 319712796 * -383503523 = 0xfe4c65f5c41d942c
Input 160 - -2074576469 * 1911074819 = 0xc8faa8de1ea6fd01
Input 161 - -1383524532 * 1796262786 = 0xdd82e3c0dd71898
Input 162 - 1807308841 * 1822349377 = 0x2db5089571bffa69
Input 163 - -1527343832 * 107667676 = 0xfdb7c5a238fb2e60
Input 164 - 524318596 * -1599424649 = 0xf45caa637a0eea5c
Input 165 - 1488670221 * 579912250 = 0xbfb0d0bace588f2
Input 166 - 747598074 * 1137527449 = 0xbcd466afe94c96a
Input 167 - 246114683 * -1878460116 = 0xf995857cbae6c424
Input 168 - 1416233495 * 546582163 = 0xabe1ce79a8b5535
Input 169 - 1003045139 * 1812063097 = 0x1939585cfbe69afb
Input 170 - 34097309 * -999465972 = 0xff86ed3d3154d35c
Input 171 - 855718216 * -218326864 = 0xfd68427e3dba2180
Input 172 - 2020576483 * -2007910601 = 0xc7b22772c98089c5
Input 173 - 1730366039 * 1480495707 = 0x238d5710b3030aed
Input 174 - -391585211 * -1977327870 = 0xabed6ca80fd5f8a
Input 175 - -1574008061 * -1617405295 = 0x235487c5a003cab3
Input 176 - -944818531 * -156268262 = 0x20c8a7b3f5890f2
Input 177 - 1009311449 * 1346895014 = 0x12ddb058dc0d1cb6
Input 178 - 1348873212 * -161999180 = 0xfcf7ac88963e9530
Input 179 - 1074161569 * -255589270 = 0xfc309f87e3896caa
Input 180 - 939492001 * 838991311 = 0xaf05707c054bd2f
Input 181 - -34167178 * -56172954 = 0x6d1912d24a704
Input 182 - 727043879 * -2042919187 = 0xeb632ee530dc311b
Input 183 - 86364705 * 1499464039 = 0x1cc14437d69ac47
Input 184 - -503480467 * 725395779 = 0xfaee77f9ab034e87
Input 185 - -750285791 * -293181479 = 0x30d7d74e84d0ef9
Input 186 - 2068205098 * -1489814648 = 0xd53d3a7fa9bb3450
Input 187 - 1022633640 * 733201444 = 0xa67cfe8bf660fa0
Input 188 - -267245308 * 2082137598 = 0xf8471f668b1d45f8
Input 189 - 2077825339 * -780979358 = 0xe97adfc029a36996
Input 190 - -227832924 * -264611492 = 0xd62ee5e55a32f0
Input 191 - -1485331162 * 1031539858 = 0xeabc9bf52a61f7ac
Input 192 - -1399865976 * 613598484 = 0xf41460654fa476a0
Input 193 - -2138792749 * -1105127231 = 0x20cd545623e10313
Input 194 - 631991117 * -1076525084 = 0xf68ee560ac522b94
Input 195 - 1185131674 * 1712280238 = 0x1c29712ce98ea4ac
Input 196 - 1203334916 * 988943901 = 0x1083d6a3e953eb74
Input 197 - -1369068157 * -1893553730 = 0x23fa11cec768423a
Input 198 - 259505549 * 1260385011 = 0x48a02b622eb36d7
Input 199 - 653863124 * 70633016 = 0xa41465fff8e660
PASSED